### PR TITLE
Implementation of mean in categorical

### PR DESCRIPTION
**Pull request by fotisdr**
> NotImplementedError: mean is not implemented: Categorical

Defined a _mean method for implementing the mean of the Categorical distribution, following a previous PR (https://github.com/tensorflow/probability/pull/1411).

### CODE REVIEW

File: tensorflow_probability/python/distributions/categorical.py [hunk 1]
```diff
@@ -333,6 +333,10 @@ def _entropy(self):
         _mul_exp(log_probs, log_probs),
         axis=-1)
 
+  def _mean(self):
+    probs = self.probs_parameter()
+    return tf.reduce_sum(tf.range(self._num_categories(probs),dtype=probs.dtype) * probs, axis=-1) / tf.reduce_sum(probs, axis=-1)
```
[the remark this answers is not in the document]
Comment (csuter) [reply tensorflow_probability/python/distributions/categorical.py:338]: sorry, meant [weighted_logsumexp](https://www.tensorflow.org/probability/api_docs/python/tfp/math/reduce_weighted_logsumexp)

[the remark this answers is not in the document]
Comment (fotisdr) [reply tensorflow_probability/python/distributions/categorical.py:338]: Should we always expect that the provided probs sum up to 1 for this method, and thus add an assertion that this is the case before computing the mean? Because other methods in the Categorical work without the sum of the probs necessarily being 1 (with validate_args=False).

File: tensorflow_probability/python/distributions/categorical.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
 from tensorflow_probability.python.internal import samplers
 from tensorflow_probability.python.internal import tensor_util
 from tensorflow_probability.python.internal import tensorshape_util
+from tensorflow_probability.python.math import reduce_weighted_logsumexp
```
Comment (csuter) [inline tensorflow_probability/python/distributions/categorical.py:33]: Please import as

```
from tensorflow_probability.python.math.generic import reduce_weighted_logsumexp
```

and add this to the "categorical" `deps` list in the adjacent BUILD file:

"//tensorflow_probability/python/math:generic"

File: tensorflow_probability/python/distributions/categorical.py [hunk 3]
```diff
@@ -333,6 +334,13 @@ def _entropy(self):
         _mul_exp(log_probs, log_probs),
         axis=-1)
 
+  def _mean(self):
+    #probs = self.probs_parameter()
```
Comment (csuter) [inline tensorflow_probability/python/distributions/categorical.py:338]: new impl looks good, thanks! please

1) remove the commented out lines
2) add a test in categorical_test.py. you can look at other categorical tests and maybe normal_test.py for a hint of how these should look. think of edge cases, like some zero prob categories, etc.

File: tensorflow_probability/python/distributions/categorical.py [hunk 4]
```diff
@@ -333,6 +334,13 @@ def _entropy(self):
...
+    #return tf.reduce_sum(tf.range(self._num_categories(probs),dtype=probs.dtype) * probs, axis=-1) / tf.reduce_sum(probs, axis=-1)
+    # Implement with logits to improve numerical stability
+    logits = self.logits_parameter()
+    return tf.math.exp(reduce_weighted_logsumexp(logits,w=tf.range(self._num_categories(logits),dtype=logits.dtype),axis=-1))
```
Comment (csuter) [inline tensorflow_probability/python/distributions/categorical.py:342]: please format as

```
return tf.math.exp(
    reduce_weighted_logsumexp(
        logits,
        w=tf.range(self._num_categories(logits), dtype=logits.dtype),
        axis=-1))
```

**Comment by csuter** [comment]
A few more comments

**Review by csuter — CHANGES_REQUESTED** [review]

**Comment by SiegeLordEx** [comment]
This got implemented in b14ae1d79de4a52d834a3ba2dc88f7e5d849e6c7.
